### ISSUE TITLE
Different sizes when combining prechipped dataset

### ISSUE THREAD

**Issue by paolo-fraccaro** [microsoft/torchgeo#737]
### Description

Hi. 

I am using a pre-chipped dataset and when I am combining image and label RasterDatasets with the IntersectionDataset there seems to be some rounding issue for which the result  is not only 512x512 anymore, but sometimes 513x513 or 1x513 or 513x1 and other combinations. This does not allow to create batches greater than 1. Can you have a look please? Many thanks.

Sample data and code attached. Full dataset available from here https://github.com/cloudtostreet/Sen1Floods11

[data_samples.zip](https://github.com/microsoft/torchgeo/files/9392147/data_samples.zip)

### Steps to reproduce

```python

from torch.utils.data import DataLoader
from torchgeo.datasets import RasterDataset, stack_samples
from torchgeo.samplers import PreChippedGeoSampler

# define datasets
label_folder = 'data_samples/LabelHand/'
ds_labels = RasterDataset(root=label_folder)
ds_labels.is_image = False

image_folder = 'data_samples/S1Hand/'
ds_inputs = RasterDataset(root=image_folder)

# combine with intersection
ds = ds_inputs & ds_labels

# create data loader
sampler = PreChippedGeoSampler(ds)
dl = DataLoader(ds, sampler=sampler, collate_fn=stack_samples)

# check shapes
for sample in dl:
    image = sample["image"]
    label = sample["mask"]
    print('Image shape: ' +  str(image.shape))
    print('Label shape: ' + str(label.shape))
    print('#########')

```

### Version

0.3.0

**Comment by adamjstewart** [comment]
Thanks for reporting this! This is most closely related to #319 and #376. Basically, rtree considers two bounding boxes to overlap even if only the edge is overlapping:
```
+---------+---------+
|         |         |
|    A    |    B    |
|         |         |
+---------+---------+
```
With the above images, there is a 512 x 0 pixel region of overlap between A and B.

There are several ways we could solve this. `PreChippedGeoSampler` was designed for use with a single curated benchmark dataset, it wasn't really designed for combining two RasterDatasets. So we could:

1. Create a single `Sen1Floods11` dataset (either `GeoDataset` or `NonGeoDataset`)
2. Use `PreChippedGeoSampler(ds_labels)` instead
3. Use `RandomGeoSampler(ds, 512, len(ds_labels))` instead
4. Skip patches with 0 area in `PreChippedGeoSampler`

With our other samplers, we can go a step further and say "skip all patches with area less than `patch_size`". But `PreChippedGeoSampler` was designed to make it easy to use curated benchmark datasets where all patches have the same size, so there is no `patch_size` parameter. But if there happen to be any patches in the dataset with any overlapping area, `PreChippedGeoSampler` won't be useful.

**Comment by paolo-fraccaro** [comment]
Thanks Adam this is most helpful. I will follow the suggested approach!
